troff/troff.d/otfdump.1
.\"
.\" Sccsid @(#)otfdump.1	1.4 (gritter) 2/2/07
.TH OTFDUMP 1 "2/2/07" "Heirloom Documentation Tools" "User Commands"
.SH NAME
otfdump \- dump the contents of OpenType font files
.SH SYNOPSIS
.HP
.ad l
.nh
\fBotfdump\fR
[\fB\-ckns\fR]
font ...
.br
.hy 1
.ad b
.SH DESCRIPTION
.I otfdump
prints the contents of OpenType or TrueType font files
in ASCII format
just as
.I troff
interprets it.
By default,
all available information is printed.
Options can be used to restrict the output:
.TP
.B \-c
prints only the characters present in the font,
in the format `\fBchar \fIstring \fBwidth \fInumber\fR'.
.TP
.B \-k
prints only the kerning pairs present in the font,
in the format `\fBkernpair \fIstring1 string2 \fBwidth \fInumber\fR'.
.TP
.B \-n
prints only the PostScript name of the font,
in the format `\fBname \fIstring\fR'.
.TP
.B \-s
prints only the substitution features available with the font,
in the format `\fBfeature \fIstring \fBsubstitution \fIstring1 string2\fR'.
.SH "SEE ALSO"
troff(1)


CHANGES
2014-10-29
    Documentation revised by Pierre-Jean Fichet:  doc/fonts/
    changed to use free available OTF fonts.

2014-10-27
    col(1) bugfix: Avoid to access chars before string.

    troff: fdprintf changed back to libc fprintf.

    troff: 32 bit bug in register and macro allocation fixed (Bug
    reported by Anthony J. Bentley).

    nroff/troff made -Wall clean.

2014-10-22
    OpenSolaris col(1) added.

2014-10-06
    Drawing of filled polygone (\D'P ...'), filled circle (\D'C
    ...') and filled ellipse (\D'E ...') added.

2014-09-21
    *roff: Option -V for version information added.

2014-09-14
    Any version of roff (e.g. UNIX V7) does count \} as a macro
    argument regarding \n(.$.  This makes \n(.$ rather random if
    you cannot force users to put \} on a separate line *or* to
    put it behind the macro call and substract 1 from \n(.$.  To
    make \n(.$ meaningful again this bug has been fixed now, i.e.
    \} is not considered as a macro argument anymore.

2014-09-07
    Documentation revised by Pierre-Jean Fichet:  doc/just/ and
    doc/quickstart/ changed to use free available OTF fonts.

2014-09-03
    Bug reported by Pierre-Jean Fichet fixed:  If character '"'
    is used in a .char request it did not work in a .if request
    as the string compare delimiter.

2014-08-28
    Bug reported by Blake McBride fixed:  Very old versions of
    the mm macros inserted cut marks in the document for
    typesetters which had used endless paper rolls.  Although
    they are only of historical significance they had never been
    removed in the Solaris version of mm.  They are now switched
    of (via .rm )k).

2014-08-23
    Bugfix by Steffen Nurpmeso:  In .rm code a function return
    value is unchecked used as a pointer.

2014-08-20
    eqn(1) did not compile because of bison to BSD yacc differ-
    ences.  Sorry, had not tested on Linux this time...

2014-08-18
    Experimental bugfix: When .bp is directly followed by .. and
    .bp triggers a macro which calls a macro then document text
    is printed at that trap position.  This bug is also in other
    ditroff derivatives (e.g. DWB and Plan9).  It is difficult to
    track the bug to it's root cause so it is an experimental fix
    at the moment.

2014-08-17
    \fP after \fS did set the font before the previous font and
    not the previous one.

2014-08-16
    Bug reported by Pierre-Jean Fichet fixed:  If character "!"
    is used in a .char request it did not work in a .if request
    as the NOT operator.

2014-08-14
    The eqn parser used type int for all token values--even
    pointers.  This may work fine on 32 bit hardware but causes
    problems on 64 bit.  A new data type had been introduced to
    separate int tokens from pointers.

    Also a bug in the parser regarding movements (e.g. FWD) had
    been fixed.

2014-07-27
    A variable had been initilized after use.  This variable
    minus a literal value is used as an array index.  When data
    from the array is read an invalid memory segment is accessed.
    This had caused crashes on at least OpenBSD.


troff/troff.d/otfdump.c
/*
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License, Version 1.0 only
 * (the "License").  You may not use this file except in compliance
 * with the License.
 *
 * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
 * or http://www.opensolaris.org/os/licensing.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */

/*
 * Copyright (c) 2005 Gunnar Ritter, Freiburg i. Br., Germany
 *
 * Sccsid @(#)otfdump.c	1.15 (gritter) 3/16/06
 */

static enum show {
	SHOW_CHARS		= 001,
	SHOW_KERNPAIRS		= 002,
	SHOW_SUBSTITUTIONS	= 004,
	SHOW_NAME		= 010
} show;

#include <stdarg.h>

static void	print(enum show, const char *, ...);

#define	DUMP
#include <stdio.h>
#include "otf.c"
#include "afm.c"
#include "dpost.d/getopt.c"

#include <libgen.h>

static const char	*progname;
struct dev	dev;
char	*chname;
short	*chtab;
int	nchtab;

static int	prname;

static void
print(enum show s, const char *fmt, ...)
{
	va_list	ap;

	if (show & s) {
		if (prname)
			printf("%s: ", filename);
		va_start(ap, fmt);
		vprintf(fmt, ap);
		va_end(ap);
		putchar('\n');
	}
}

void
verrprint(const char *s, va_list ap)
{
	fprintf(stderr, "%s: ", progname);
	vfprintf(stderr, s, ap);
	putc('\n', stderr);
}

void
errprint(const char *s, ...)
{
	va_list	ap;

	va_start(ap, s);
	verrprint(s, ap);
	va_end(ap);
}

static void
devinit(void)
{
	dev.res = 72000;
	dev.hor = dev.vert = 1;
	dev.unitwidth = 1;
	dev.afmfonts = 1;
}

static void
usage(void)
{
	errprint("usage: %s [-ckns] font ...", progname);
	exit(2);
}

static int
dump(const char *name)
{
	struct afmtab	A;
	struct stat	st;
	FILE	*fp;
	char	*cp;

	if ((fp = fopen(filename = name, "r")) == NULL) {
		errprint("%s: cannot open", filename);
		return 1;
	}
	memset(&A, 0, sizeof A);
	a = &A;
	a->file = a->path = (char *)filename;
	a->base = malloc(strlen(filename) + 1);
	strcpy(a->base, filename);
	a->base = basename(a->base);
	if ((cp = strrchr(a->base, '.')) != NULL)
		*cp = '\0';
	if (fstat(fileno(fp), &st) < 0) {
		errprint("%s: cannot stat", filename);
		return 1;
	}
	size = st.st_size;
	contents = malloc(size);
	if (fread(contents, 1, size, fp) != size) {
		errprint("%s: cannot read", filename);
		return 1;
	}
	fclose(fp);
	return otfget(a, contents, size) != 0;
}

int
main(int argc, char **argv)
{
	int	i, e = 0;

	progname = basename(argv[0]);
	devinit();
	while ((i = getopt(argc, argv, "ckns")) != EOF) {
		switch (i) {
		case 'c':
			show |= SHOW_CHARS;
			break;
		case 'k':
			show |= SHOW_KERNPAIRS;
			break;
		case 'n':
			show |= SHOW_NAME;
			break;
		case 's':
			show |= SHOW_SUBSTITUTIONS;
			break;
		default:
			usage();
		}
	}
	if (show == 0)
		show = 0xFFFFFFFF;
	if (argc < optind + 1)
		usage();
	prname = argc > optind + 1;
	for (i = optind; i < argc; i++)
		e |= dump(argv[i]);
	return e;
}

void
afmaddchar(struct afmtab *a, int C, int tp, int cl, int WX, int B[4], char *N,
		enum spec s, int gid)
{
	if (N)
		print(SHOW_CHARS, "char %s width %d", N, _unitconv(WX));
}

void
afmalloc(struct afmtab *a, int n)
{
}

void
afmaddkernpair(struct afmtab *a, int ch1, int ch2, int k)
{
}

static void
kernpair(int first, int second, int x)
{
	char	*s1, *s2;

	if (x) {
		s1 = GID2SID(first);
		s2 = GID2SID(second);
		if (s1 && s2)
			print(SHOW_KERNPAIRS, "kernpair %s %s width %d",
				s1, s2, _unitconv(x));
	}
}

static void
kerninit(void)
{
}

static void
kernfinish(void)
{
}
